grudge/reductions.py

```diff
@@ -305,6 +319,92 @@ def elementwise_sum(dcoll: DiscretizationCollection, *args) -> DOFArray:
     )
 
 
+def elementwise_max(dcoll: DiscretizationCollection, *args) -> DOFArray:
+    r"""Returns a vector of DOFs with all entries on each element set
+    to the maximum over all DOFs on that element.
+
+    May be called with ``(dcoll, vec)`` or ``(dcoll, dd, vec)``.
+
+    :arg dcoll: a :class:`grudge.discretization.DiscretizationCollection`.
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization if not provided.
+    :arg vec: a :class:`~meshmode.dof_array.DOFArray`
+    :returns: a :class:`~meshmode.dof_array.DOFArray` whose entries
+        denote the element-wise max of *vec*.
+    """
+
+    if len(args) == 1:
+        vec, = args
+        dd = dof_desc.DOFDesc("vol", dof_desc.DISCR_TAG_BASE)
+    elif len(args) == 2:
+        dd, vec = args
+    else:
+        raise TypeError("invalid number of arguments")
+
+    dd = dof_desc.as_dofdesc(dd)
+
+    if isinstance(vec, np.ndarray):
+        return obj_array_vectorize(
+            lambda vi: elementwise_max(dcoll, dd, vi), vec
+        )
+
+    actx = vec.array_context
+
+    return DOFArray(
+        actx,
+        tuple(
+            actx.call_loopy(
+                _map_elementwise_reduction(actx, "max"),
+                operand=vec_i
+            )["result"]
+            for vec_i in vec
+        )
+    )
```

Comment:
Why not share all this code between `sum`/`min`/max`? It's super repetitive, the only difference is the string of the operation name.

Reply:
@inducer How's this: https://github.com/inducer/grudge/pull/125/commits/a2b5c49b1292d12ed99fe22d03a6f2df0a027ea6?